### ISSUE TITLE
Custom serializer for enums

### ISSUE THREAD

**Issue by aleien** [Kotlin/kotlinx.serialization#1344]
Hello!

I want to implement custom serializer for enums with fallback, because currently kotlinx fails to convert enum when backend sends new type, that is not supported yet. I wanted to use StringDesriptor, but it recently gone from api ): What is correct way now to implement custom enum serializer?

**Comment by wman1980** [comment]
One way to handle that with a custom serializer would be:

```
@Serializable(with = DrmTypeSerializer::class)
enum class DrmType(val key: String) {
    Widevine("widevine"),
    None("none");

    companion object {
        fun findByKey(key: String, default: DrmType = DrmType.None): DrmType {
            return DrmType.values().find { it.key == key } ?: default
        }
    }

}

@Serializer(forClass = DrmType::class)
object DrmTypeSerializer : KSerializer<DrmType> {
    override val descriptor: SerialDescriptor =
        PrimitiveSerialDescriptor("DrmType", PrimitiveKind.STRING)

    override fun serialize(encoder: Encoder, value: DrmType) {
        encoder.encodeString(value.key)
    }

    override fun deserialize(decoder: Decoder): DrmType {
        return try {
            val key = decoder.decodeString()
            DrmType.findByKey(key)
        } catch (e: IllegalArgumentException) {
            DrmType.None
        }
    }
}
```

Unfortunately this is very specific and I would like to have a more generic approach.

I found this old discussion https://github.com/Kotlin/kotlinx.serialization/issues/31, but its a bit hard to follow and it seems to be outdated because I can not find any `CommonEnumSerializer `.

@sandwwraith can you give me some hint how enums could be handled with default values to be forward compatible?

Thanks

**Comment by sandwwraith** [comment]
@wman1980 We now have `coerceInputValues` key in Json that allows using default value for unknown enum element. Is this sufficient for your case? If not, custom serializer you mentioned should be sufficient. What do you mean by 'more generic'?

**Comment by wman1980** [comment]
can be closed

**Comment by himanshufoodpanda** [comment]
But if you use coerceInputValues = true then you can't have any Optional<> wrapper type over your primitive data type to check if values are there in json or not.

**Comment by tadfisher** [comment]
`coerceInputValues` also does not work when decoding a collection of enum values, e.g. `List<EnumType>`. This caught us in production today.

What would be very nice is a `@SerialFallback` annotation or similar that we could use at the enum declaration, rather than relying on default parameters in the use sites.

**Comment by silverhammermba** [comment]
I found a way to do this without too much code duplication using a generic custom serializer https://stackoverflow.com/a/69438279/503402

**Comment by sandwwraith** [comment]
`coerceInputValues` flag works in general case, for lists, there is a separate ticket: #1113
